### PR TITLE
Add sort command

### LINKED ISSUE #11: As a user, I can sort my contacts
so that I can view my contacts in a more convenient manner and get the contacts I want quickly

### LINKED ISSUE #72: Implement sorting function
Permanent sort of the address book by name, address and primary tag

### PR DESCRIPTION
**Pull request by bangyiwu**
First implementation of sorting, sort 1 allows sorting by name, sort 2 allows sorting by address and sort 3 allows sorting by the person's first tag

### CODE REVIEW

**Comment by bangyiwu** [comment]
I will fix the cli fails before reopening

**Comment by LinkedInk** [comment]
Model should not handle the sortPerson of the addressbook. Should consider moving the function into the excute part of SortCommand and the class in general. Model class has getAddressBook and setAddressBook to help with that.

**Comment by chan-j-d** [comment]
Is the sort method supposed to sort the current shown list? Or is it supposed to sort the full list of contacts before showing it? Because the former would be much easier to implement and I believe is what we intended to do originally? There is the `model.getFilteredPersonList()` method which gives the last shown list.

Also, I agree with the Hendey's comment about the sortPerson method. I do not think that all the model classes need to have such a method.

**Comment by bangyiwu** [comment]
Challenges faced: so far only able to sort the address book permanently, if anyone has an idea on how we can sort the address book temporarily, let me know

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 1]
```diff
@@ -11,6 +11,7 @@
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
+    public static final Prefix PREFIX_SORT = new Prefix("s/");
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java]: Is the prefix necessary at all?

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+     * and returns a {@code SortCommand} object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public SortCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_SORT);
+
+        Index index;
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/parser/SortCommandParser.java]: Should be able to obtain the index directly similar to the Delete command for example

File: src/main/java/seedu/address/logic/commands/SortCommand.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+public class SortCommand extends Command {
+
+    public static final String COMMAND_WORD = "sort";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Sorts the currently displayed list"
+            + "by the index command entered "
+            + "1 will be sort by date added\n"
+            + "2 will be sort by alphabetical order of their names\n"
+            + "3 will be sort by alphabetical order of their address\n"
+            + "4 will be sort by alphabetical order of their first tag\n"
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/commands/SortCommand.java]: This seems to be outdated compared to what is actually implemented

File: src/main/java/seedu/address/logic/commands/SortCommand.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
+package seedu.address.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_SORT;
+
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/commands/SortCommand.java:12]: Would be nice to have a brief Javadocs description of the class

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_SORT;
+
+import seedu.address.commons.core.index.Index;
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.logic.commands.SortCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+/**
+ * Parses input arguments and creates a new {@code RemarkCommand} object
```
Comment (chan-j-d) [inline src/main/java/seedu/address/logic/parser/SortCommandParser.java]: This Javadocs is outdated

File: src/main/java/seedu/address/model/AddressBook.java [hunk 6]
```diff
@@ -95,6 +96,10 @@ public void removePerson(Person key) {
 
     //// util methods
 
+    public void sortPerson(Comparator<Person> c) {
+        persons.sortPersons(c);
+    }
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/AddressBook.java:101]: Should probably remove this as it is not used at all

File: src/main/java/seedu/address/model/Model.java [hunk 7]
```diff
@@ -124,6 +125,7 @@
      */
     void updateFilteredPersonList(Predicate<Person> predicate);
 
+    void sortPerson(Index index);
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/Model.java]: Should remove this from the interface and the implementing classes

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 8]
```diff
@@ -97,13 +98,18 @@ public void setPersons(List<Person> persons) {
         internalList.setAll(persons);
     }
 
+    public void sortPersons(Comparator<Person> c) {
+        internalList.sort(c);
+    }
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/person/UniquePersonList.java:103]: Should probably remove this as it is only used by the unused sortPersons method in AddressBook

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 9]
```diff
@@ -176,7 +177,12 @@ public void setPerson(Person target, Person editedPerson) {
         }
 
         @Override
-        public ObservableList<Person> getFilteredPersonList() {
+        public void sortPerson(Index index) {
+            throw new AssertionError("This method should not be called.");
+        }
```
Comment (chan-j-d) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java]: Should probably remove this as it is not used at all

File: src/main/java/seedu/address/model/ModelManager.java [hunk 10]
```diff
@@ -156,15 +163,58 @@ public void setPerson(Person target, Person editedPerson) {
         addressBook.setPerson(target, editedPerson);
     }
 
+    @Override
+    public void sortPerson(Index index) {
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/ModelManager.java]: This method should probably take in a Comparator instead in line with how the `updateFilteredPersonList` method works. Model Manager shouldn't need to know exactly which Comparators it needs

File: src/main/java/seedu/address/model/ModelManager.java [hunk 11]
```diff
@@ -156,15 +163,58 @@ public void setPerson(Person target, Person editedPerson) {
...
+                if (o1.getTags().size() != 0 && o2.getTags().size() == 0) {
+                    return -1;
+                }
+                if (o1.getTags().size() == 0 && o2.getTags().size() == 0) {
+                    return o1.getName().fullName.compareToIgnoreCase(o2.getName().fullName);
+                }
+                return o1.getTags().iterator().next().tagName
+                        .compareToIgnoreCase(o2.getTags().iterator().next().tagName);
+            }
+        };
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/ModelManager.java]: The three comparators should probably declared as static constants within the `SortCommand` class itself?

File: src/main/java/seedu/address/model/ModelManager.java [hunk 12]
```diff
@@ -156,15 +163,58 @@ public void setPerson(Person target, Person editedPerson) {
...
+        int indexNumber = index.getOneBased();
+        if (indexNumber == 1) {
+            sortedPersons.comparatorProperty().setValue(nameComparator);
+        } else if (indexNumber == 2) {
+            sortedPersons.comparatorProperty().setValue(addressComparator);
+        } else if (indexNumber == 3) {
+            sortedPersons.comparatorProperty().setValue(tagComparator);
+        }
```
Comment (chan-j-d) [inline src/main/java/seedu/address/model/ModelManager.java]: This should probably be done in the SortCommand class itself. `execute` chooses the appropriate Comparator in the execute command to pass to `ModelManager`.

**Review by chan-j-d — CHANGES_REQUESTED** [review]
Please take a look at the suggestions

**Review by chan-j-d — APPROVED** [review]
LGTM!

**Pull request merged**